sdk/core/azure-core-amqp/inc/azure/core/amqp/internal/common/runtime_context.hpp

```diff
@@ -105,5 +110,29 @@ namespace Azure { namespace Core { namespace Amqp { namespace Common { namespace
     Azure::Core::Context m_context;
   };
 
+  /**
+   * Invoke a Rust AMQP builder API, checking for error.
+   *
+   * @param api - Flat C API to invoke. The first parameter MUST be a RustCallContext object, the
+   * second parameter must be a Rust client object.
+   * @param builder - Unique Pointer to a Rust builder object.
+   * @param args - remaining arguments to the flat C API.
+   *
+   * This function will check the return from the API, and if it is null, will throw an exception
+   * with error information from the callContext.
+   *
+   */
+  template <typename Api, typename T, typename... Args>
+  void invoke_builder_api(Api& api, T& builder, Args&&... args)
```

Comment:
Should this function naming be PascalCased, i.e. `InvokeBuilderApi`?

Reply:
Another thought - do you think if we make `builder` first parameter, and `api` a second parameter, then `InvokeBuilderApi(builder, api, arg)`, becomes more readable as `builder.api(arg)`?

Reply:
The decision to make the builder the second parameter was intentional - that way the invocation of the API is consistent with the actual builder API - so when you want to call:

```cpp
session_set_max_bytes(builder, 37);
```

there are two choices in experience:
```cpp
invoke_builder_api(rustBuilder, session_set_max_bytes, 37);
```

or
```cpp
invoke_builder_api(session_set_max_bytes, rustBuilder, 37);
```

The second invocation is dramatically closer to the rust API shape.

The C++17 version of this incarnation is even more interesting - that one looks like:

```cpp
invoke_builder_api<session_set_max_bytes>(rustBuilder, 37);
```

and I would *much* prefer to use that invocation but...

Reply:
Renamed to be camelCase - I originally made it snake_case the rust APIs are all snake case (the standard for C and Rust APIs), but your point is valid.

Reply:
It might be possible - I cheated with declaring it as `auto`, which is C++17, but this could mean that there is ultimately a declaration that would fit, which we could use in C++14 to declare more explicitly - https://godbolt.org/z/fhYcYeGEz

Reply:
Unfotunately that solution doesn't compile with MSVC :(. 

FWIW - this was the solution that STL came up with when I asked about this problem in the context of OpenSSL APIs about 3 years ago - I've modified their solution a bit, but I suspect it's the only one which works reliably across compilers :(.